rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+requires a goodly chunk of the code base to be adjusted. This is fine when work
+is demonstrated to have a positive effect but is wasted time when the change is
+neutral or negative. As well, if a user desires a vector build with debug
+symbols present this imposes a severe size constraint on the user _unless_ they
+carefully curate among our current zoo of feature flags, or other custom build
+goals. It is our belief that by extracting a vector "core" we can:
+
+* reduce iteration costs in engineering work,
+* reduce the burden for casual contributors,
+* enable more focused correctness and performance tests and
+* reduce experimentation costs, pushing toward a fail-fast model of work.
+
+## Motivation
+
+We want to push the vector project into new, more competitive spaces. This will
+mean integrating with more sources of data, allowing our users to do new things
+with their in-flight data and also consistently, release after release, use less
+resources to do it. As noted in the summary build times are long and growing,
+imposing a burden on development. It is _hard_ to benchmark and improve your
+code if the feedback cycle is 30 minutes. Our flat structure raises the barrier
+to entry for contributions; users on discord have referred to our compile times
+as "punishing". If a user contributes a new sink, say, then they are responsible
+for:
+
+* writing the sink and its correctness tests,
+* writing benchmarks,
+* resolving any incompatibilities between their new code and the existing vector
+  model and
+* waiting for CI to approve their changes.
+
+With regard to benchmarks we generally do not request contributors to add them
+as doing so is too complicated. Our flat structure promotes a conflation between
+unit and integrated benchmark tests, not to mention you need a fair bit of
+memory to _link_ vector to run these tests yourself. The feedback loop here is
+hours long. Any incompatibilities between new and existing code are quite
+time-consuming to resolve as our flat structure encourages a "big bang" mindset
+to changes, if not mandates it. The flatness of our project structure does mean,
+however, that relatively simple sources/transforms/sinks are convenient to
+introduce, there is "one place" for such things.
+
+The vector project does have some sub-packages. VRL -- in `lib/vrl` -- is one
+such. The development experience here has a tighter feedback loop than in vector
+proper. VRL benchmarks can be built independent of vector -- avoiding its link
+cost -- and its tests end up clearly focused on VRL itself. This _does_ mean
+that integration tests remain to be written in the top-level of the project but
+the overall experience is improved. Changes in VRL are isolated to VRL, though
+experience with `Lookup`/`LookupBuf` -- see
+[5374](https://github.com/timberio/vector/pull/5374) -- does suggest that
+aggressive integration of a cross-cutting concern will make things challenging.
```

Comment:
Yeah, I think this was a good illustration of what can go wrong when trying to draw seams early on. It wasn't entirely clear how to separate concerns between VRL and vector proper, so we ended up with some very similar-but-not-quite-the-same implementations on either side, a somewhat ambiguous "shared" crate in between, etc.